integrated_channels/canvas/views.py

```diff
@@ -0,0 +1,112 @@
+"""
+Views containing APIs for cornerstone integrated channel
+"""
+
+from django.conf import settings
+from django.apps import apps
+import requests
+from rest_framework.response import Response
+from rest_framework import generics, status
+from six.moves.urllib.parse import urljoin
+
+from enterprise.utils import get_enterprise_customer
+from integrated_channels.canvas.models import CanvasEnterpriseCustomerConfiguration
+class CanvasCompleteOAuthView(generics.ListAPIView):
+    """
+        **Use Cases**
+
+            Retrieve and save a Canvas OAuth refresh token after an enterprise customer
+            authorizes to integrated courses.
+
+        **Example Requests**
+
+            GET /canvas/oauth-complete?code=123abc&state=abc123
+
+        **Query Parameters for GET**
+
+            * code: The one time use string generated by the Canvas API used to fetch the
+            access and refresh tokens for integrating with Canvas.
+
+            * state: The user's enterprise customer uuid used to associate the incoming
+            code with an enterprise configuration model.
+
+        **Response Values**
+
+            If the request for information about the course list is successful, an HTTP 200 "OK" response
+            is returned.
+
+            If code is not provided, a 400 error is returned.
+
+            If state is not provided, a 400 error is returned.
+
+            If the specified state is not valid or contained in the set of registered enterprise customers
+            a 404 error is returned.
+
+    """
+    def get(self, request, *args, **kwargs):  # pylint: disable=arguments-differ
+        app_config = apps.get_model('canvas', 'CanvasGlobalConfiguration').current()
+        canvas_oauth_token_path = app_config.oauth_token_auth_path
+        client_code = request.GET.get('code')
+        enterprise_customer_uuid = request.GET.get('state')
+        if not enterprise_customer_uuid:
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data={
+                    "message": (
+                        u"Enterprise ID required to integrate with Canvas."
+                    )
+                })
+
+        if not client_code:
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data={
+                    "message": (
+                        u"Canvas code token is required to complete the Oauth process."
+                    )
+                })
+
+        enterprise_customer = get_enterprise_customer(enterprise_customer_uuid)
```

Comment:
I think there is a get_object_or_404() method in django_http module that can make this more concise (I do not know if it allows custom messages ... ) worth a check

Reply:
I don't think it does but the 404 raises, so wherever the heck we want to actually submit a custom message (if we want a custom message) we can.

Reply:
> Got this error on: switching to this branch and reinstalling into edx-platform and restarting lms, then visiting: http://localhost:18000/canvas/oauth-complete
> 
> ```
> AttributeError at /canvas/oauth-complete
> 'CanvasGlobalConfiguration' object has no attribute 'oauth_token_auth_path'
> ... 
> /edx/src/edx-enterprise/integrated_channels/canvas/views.py in get
>         canvas_oauth_token_path = app_config.oauth_token_auth_path 
> ```

Talked offline about this error- the oauth token path was being pulled from the model instead of the app config, where it actually lives. Updating to the correct source

Reply:
looking at the docks, it seems like `get_object_or_404()` is just calling 
```
    try:
        obj = MyModel.objects.get(<param query>)
    except <param model>.DoesNotExist:
        raise Http404("No <param model> matches the given query.")
```
So I'll just implement my own version of this with an understandable message

Reply:
but also worth noting that the original `get_enterprise_customer()` is essentially doing the same thing.  ¯\_(ツ)_/¯

Reply:
Actually, `get_enterprise_customer()` returns none on `except EnterpriseCustomer.DoesNotExist:`, so why not use the method and throw an error if we don't have an enterprise customer afterwards?

Reply:
Yeah the get_object_or_404 wrapper is because these were throwing 400 or 500's an exception traces for what amounted to "the provided uuid just doesn't exist/can't be found". So it may or may not be necessary to wrap and if your submethod is already wrapping "object cannot be found" results than it's fine.